website/addons/googledrive/model.py

```diff
@@ -2,226 +2,200 @@
-    refresh_token = fields.StringField()
-    expires_at = fields.DateTimeField()
+    auth_url_base = '{}{}'.format(drive_settings.OAUTH_BASE_URL, 'auth?access_type=offline&approval_prompt=force')
+    callback_url = '{}{}'.format(drive_settings.API_BASE_URL, 'oauth2/v3/token')
+
+    default_scopes = drive_settings.OAUTH_SCOPE
+    _auth_client = GoogleAuthClient()
+    _drive_client = GoogleDriveClient()
+
+    def handle_callback(self, response):
+        client = self._auth_client
+        info = client.userinfo(response['access_token'])
+        return {
+            'provider_id': info['sub'],
+            'display_name': info['name'],
+            'profile_url': info.get('profile', None)
+        }
+
+    def _refresh_token(self, access_token, refresh_token):
+        """ Handles the actual request to refresh tokens
+
+        :param str access_token: Access token (oauth key) associated with this account
+        :param str refresh_token: Refresh token used to request a new access token
+        :return dict token: New set of tokens
+        """
+        client = self._auth_client
+        if refresh_token:
+            token = client.refresh(access_token, refresh_token)
+            return token
+        else:
+            return False
 
     def fetch_access_token(self, force_refresh=False):
         self.refresh_access_token(force=force_refresh)
-        return self.access_token
+        return self.account.oauth_key
 
     def refresh_access_token(self, force=False):
-        if self._needs_refresh() or force:
-            client = GoogleAuthClient()
-            token = client.refresh(self.access_token, self.refresh_token)
-
-            self.access_token = token['access_token']
-            self.refresh_token = token['refresh_token']
-            self.expires_at = datetime.utcfromtimestamp(token['expires_at'])
-            self.save()
-
-    def revoke_access_token(self):
-        # if there is only one osf user linked to this google drive user oauth, revoke the token,
-        # otherwise, disconnect the osf user from the googledriveoauthsettings
-        if len(self.googledriveusersettings__accessed) <= 1:
-            client = GoogleAuthClient()
-            try:
-                client.revoke(self.access_token)
-            except:
-                # no need to fail, revoke is opportunistic
-                pass
+        """ If the token has expired or will soon, handles refreshing and the storage of new tokens
 
-            # remove the object as its the last instance.
-            GoogleDriveOAuthSettings.remove_one(self)
+        :param bool force: Indicates whether or not to force the refreshing process, for the purpose of ensuring that authorization has not been unexpectedly removed.
+        """
+        if self._needs_refresh() or force:
+            token = self._refresh_token(self.account.oauth_key, self.account.refresh_token)
+            self.account.oauth_key = token['access_token']
```

Comment:
A `refresh_token` (in this case, `self.account.refresh_token`) is passed in, so it won't.